### ISSUE TITLE
Can we replace `Array = Any` with `Array = ndarray` now?

### ISSUE THREAD

**Issue by YouJiacheng** [google/jax#10186]
It seems that problems in #943 have been resolved?

**Comment by jakevdp** [comment]
I don't think it's as simple as that – for example, inputs to JAX functions will accept any array-like object, including Python scalar types, numpy scalar types, numpy arrays, jax arrays/tracers, plus for many functions, arbitrary objects that define a `__jax_array__` method. So we'd need to account for those too, which makes it more complicated than simply setting `Array = jnp.ndarray`.

**Comment by jakevdp** [comment]
I personally prefer `Array = Any` because its false positive rate is zero 😁

**Comment by YouJiacheng** [comment]
Oh, I forget `__jax_array__`.
However, IIUC, we can still annotate the return type as `ndarray`, and its false positive rate is zero as well.
And return type annotation is very helpful for IDE auto-completing and semantic highlighting

**Comment by YouJiacheng** [comment]
@jakevdp WDYT about annotate the return type as `ndarray`, or even `Union[ndarray, Any]` (IIUC `Union[ndarray, PRNGKeyArray]` is enough), which can be very useful for development with IDE/vscode.

**Comment by jakevdp** [comment]
Sure, we could try it, and just be ready to roll it back quickly if it breaks things.

**Comment by jakevdp** [comment]
I have to say though, I have a knee-jerk dislike for using `Union[Any, ndarray]` when the result is an array. What it tells me is that the static typing system is not actually expressive enough to statically type our API, and if that's necessary I'd prefer to continue to just use `Any`.

**Comment by YouJiacheng** [comment]
@jakevdp 
I have an idea: use overload to provide a fully annotated version for the most usual case, and a `Any` annotated version for corner case.

**Comment by NeilGirdhar** [comment]
What's Jax's long term plan with this?  Perhaps the [Array API will provide annotations](https://github.com/data-apis/consortium-feedback/discussions/8) that can be used in Jax?

**Comment by jakevdp** [comment]
I wrote a design doc a while ago regarding how the project should approach type annotations... it got lost in the shuffle. I'll dust it off and try to generate some discussion.

**Comment by YouJiacheng** [comment]
❤️Does the design doc also discuss things mentioned in #10322 ?

**Comment by YouJiacheng** [comment]
BTW, now I know `__jax_array__` can be solved by `Protocol`, i.e. `Union[ndarray, JAXArray]` where `JAXArray` is a protocol implemented `__jax_array__(self) -> ndarray`.

**Comment by jakevdp** [comment]
It's still unclear whether `__jax_array__` is something we want to support broadly; see the discussion in #10065.

**Comment by jakevdp** [comment]
The design doc is coming together here: https://github.com/google/jax/pull/11859

In particular, it does answer the main question of this Issue

> Can we replace `Array = Any` with `Array = ndarray` now?

The answer is No, because we need inputs to functions to be far more flexible than `ndarray`. Inputs should be `ArrayLike`, where this accepts `jax.numpy.ndarray`, `numpy.ndarray`, numpy scalars, Python scalars, objects with `__array__` attributes, etc.

For return values, we can annotate with `jnp.ndarray`, or perhaps a new `jax.typing.NDArray`.

**Comment by NeilGirdhar** [comment]
> The answer is No, because we need inputs to functions to be far more flexible than ndarray. Inputs should be ArrayLike, where this accepts jax.numpy.ndarray, numpy.ndarray, numpy scalars, Python scalars, objects with __array__ attributes, etc.

Would it be possible to create a type alias for `jax.numpy.ArrayLike` like the [one in Numpy](https://numpy.org/devdocs/reference/typing.html#numpy.typing.ArrayLike).

**Comment by YouJiacheng** [comment]
actually in jax, the problem is simpler than numpy, since jax.numpy is not designed to be compatible with sequence, e.g. list. It is not too bad to use an union of scalar, ndarray and supportsjaxarray.

**Comment by jakevdp** [comment]
> Would it be possible to create a type alias for jax.numpy.typing.ArrayLike like the [one in Numpy](https://numpy.org/devdocs/reference/typing.html#numpy.typing.ArrayLike)?

Yes, the draft doc I linked to mentions doing that in the roadmap.

**Comment by jakevdp** [comment]
I'm going to close this issue, because I think the work going forward is better outlined and tracked elsewhere. If you have thoughts about the type annotation roadmap, please feel free to comment on https://github.com/google/jax/pull/11859
